### PR TITLE
[Feature][task-plugins] aws datasync task plugins

### LINKED ISSUE #11663: [Feature][TASK-PLUGIN] AWS DataSync Operator
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Create a new task type to deal with AWS AWS DataSync Operator, the capability of this task is to call AWS SDK APIs to do data (files/objects) sync.
Maybe support 
1. from on-premises storage and AWS
2. AWS storage services

### Use case

use this task to call AWS DataSync functions

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Tianqi-Dotes**
## Purpose of the pull request

close https://github.com/apache/dolphinscheduler/issues/11663

## Brief change log

aws datasync task plugins

## Verify this pull request
Manually verified the change by testing locally.

### CODE REVIEW

File: docs/docs/en/guide/task/datasync.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+- Click `Project -> Management-Project -> Name-Workflow Definition`, and click the "Create Workflow" button to enter the
+  DAG editing page.
+- Drag from the toolbar <img src="../../../../img/tasks/icons/datasync.png" width="15"/> task node to canvas.
+
+## Task Example
+
+First, introduce some general parameters of DolphinScheduler:
+
+- **Node name**: The node name in a workflow definition is unique.
```
Comment (caishunfeng) [inline docs/docs/en/guide/task/datasync.md]: use base doc, see https://dolphinscheduler.apache.org/en-us/docs/dev/user_doc/guide/task/appendix.html

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/pom.xml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-dms</artifactId>
+            <version>1.12.297</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.dolphinscheduler</groupId>
+            <artifactId>dolphinscheduler-common</artifactId>
+            <scope>provided</scope>
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/pom.xml:50]: Why import the ds common module?

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncHook.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                return true;
+            } else if (stopStatusSet.contains(status)) {
+                break;
+            }
+        }
+        logger.warn("double check error");
+        return false;
+    }
+
+    public Boolean doubleCheckExecStatus(TaskExecutionStatus exceptStatus, TaskExecutionStatus[] stopStatus) {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncHook.java:171]: What's the meaning of `double check`?

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    @Override
+    public void cancelApplication() throws TaskException {
+        hook.cancelDatasyncTask();
+    }
+
+
+    @Override
+    public void trackApplicationStatus() throws TaskException {
+        Boolean isFinishedSuccessfully = hook.doubleCheckFinishStatus(TaskExecutionStatus.SUCCESS, DatasyncHook.doneStatus);
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java]: Should check applicationId before check status.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            setExitStatusCode(exitStatusCode);
+        } catch (Exception e) {
+            setExitStatusCode(TaskConstants.EXIT_CODE_FAILURE);
+            throw new TaskException("datasync task error", e);
+        }
+    }
+
+    @Override
+    public void cancelApplication() throws TaskException {
+        hook.cancelDatasyncTask();
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java:111]: Should check applicationId before.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/test/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTaskTest.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        datasyncHookMockedStatic = mockStatic(DatasyncHook.class);
+        when(DatasyncHook.createClient()).thenReturn(client);
+
+        DatasyncParameters DatasyncParameters = new DatasyncParameters();
+        datasyncTask = initTask(DatasyncParameters);
+        MemberModifier.field(DatasyncTask.class, "hook").set(datasyncTask, datasyncHook);
+    }
+
+    @Test
+    public void testCreateTaskJson() {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/test/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTaskTest.java:88]: What is the purpose of this test case?  If just test the value from json, it seems meanless.

**Review by caishunfeng — CHANGES_REQUESTED** [review]

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncHook.java (hunk 3 above)
Comment (Tianqi-Dotes) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncHook.java:171]: this method double check the object status by calling query status API.
eg: when you call createTask method, you expect an `AVAILABLE` status, so you call this method to get the task status and compare the expectation

File: docs/docs/en/guide/task/datasync.md (hunk 1 above)
Comment (Tianqi-Dotes) [reply docs/docs/en/guide/task/datasync.md]: done

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/pom.xml (hunk 2 above)
Comment (Tianqi-Dotes) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/pom.xml:50]: use this method: BeanUtils.copyProperties

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/test/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTaskTest.java (hunk 6 above)
Comment (Tianqi-Dotes) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/test/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTaskTest.java:88]: the purpose is to confirm the transformation of  `using JSON data to construct data sync task`, and validate the way to convert JSON data to an entity.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java (hunk 4 above)
Comment (Tianqi-Dotes) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java]: done

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java (hunk 5 above)
Comment (Tianqi-Dotes) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java:111]: done

Comment (caishunfeng) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java:111]: It should try to get application id from taskContext if applicationId is null.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/pom.xml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        <!-- https://mvnrepository.com/artifact/com.amazonaws/aws-java-sdk-datasync -->
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-datasync</artifactId>
+            <version>1.12.289</version>
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-dms</artifactId>
+            <version>1.12.297</version>
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/pom.xml]: Don't manage the versions here, move them to the bom

Comment (Tianqi-Dotes) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/pom.xml]: done

File: dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java (hunk 5 above)
Comment (Tianqi-Dotes) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-datasync/src/main/java/org/apache/dolphinscheduler/plugin/task/datasync/DatasyncTask.java:111]: done

**Review by caishunfeng — APPROVED** [review]
LGTM of the backend.

File: dolphinscheduler-bom/pom.xml [hunk 8]
```diff
@@ -748,6 +749,102 @@
                 <artifactId>snakeyaml</artifactId>
                 <version>${snakeyaml.version}</version>
             </dependency>
+            <!-- https://mvnrepository.com/artifact/software.amazon.awssdk/datasync -->
+            <dependency>
+                <groupId>software.amazon.awssdk</groupId>
+                <artifactId>datasync</artifactId>
+                <version>${datasync.version}</version>
+               <!-- <exclusions>
```
Comment (caishunfeng) [inline dolphinscheduler-bom/pom.xml:757]: remove if no need.

Comment (kezhenxu94) [reply dolphinscheduler-bom/pom.xml:757]: This comment is not addressed

**Review by kezhenxu94 — COMMENTED** [review]
Comments are not addressed, UI part is not reviewed.

File: dolphinscheduler-bom/pom.xml (hunk 8 above)
Comment (Tianqi-Dotes) [reply dolphinscheduler-bom/pom.xml:757]: sorry, the page not refreshed and the comment not show when i merge this. have fixed this in this pr.
https://github.com/apache/dolphinscheduler/pull/12273

**Review by Amy0104 — COMMENTED** [review]
LGTM for the front end part.

**Pull request merged**